feat(cfn): request to delete CFN changeset if empty

xavileon opened the pull request:
This patch requests the deletion of the CFN change set if it's empty.
The rationale behind it is that whenever a change set does not contain
changes to the CFN template, the AWS API marks the changeset as FAILED.
However, in a CD world, an empty change should not be considered a
failure, and instead continue the normal execution flow.

In this patch we detect that the change set is empty by the status
message and request it's deletion accordingly. This is implemented as
a separate task so it can be reused in different scenarios planned for
the future (e.g. configuration option to delete all changesets that
have been executed).

Depends on https://github.com/spinnaker/clouddriver/pull/4101

xavileon commented:
@spinnaker/reviewers PTAL

marchello2000 commented on orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/providers/aws/cloudformation/DeleteCloudFormationChangeSetTask.java:
```diff
@@ -0,0 +1,79 @@
+  public TaskResult execute(@Nonnull Stage stage) {
+    String cloudProvider = getCloudProvider(stage);
+
+    Map<String, Object> task = new HashMap<>(stage.getContext());
```
`stage.getContext()` already returns a `Map`, why do you need to copy it?
If you do need to do the copy, consider renaming `task` to `stageData` - that's what we use pretty much everywhere else and I think it's more representative of what this is

xavileon replied:
Good point. We made a copy of the context map because we were using the contents of the context to create the contents of the kato task operation. In that sense, both the context and task are coupled. I will make it more explicit by creating the task object from scratch and so we don't rely on assumptions.

marchello2000 commented on orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/providers/aws/cloudformation/DeleteCloudFormationChangeSetTask.java:
```diff
@@ -0,0 +1,79 @@
+          .first();
+      return TaskResult.builder(ExecutionStatus.SUCCEEDED).build();
+    } else {
+      log.debug(
```
`debug` or `info`? Not sure how frequent/expected this logline would be (don't know much about cloudformation).

xavileon replied:
I'd leave it at `debug` for now. Right now it would be printed whenever there's a change in the cloudformation stack that is being deployed. I don't think it has operational importance (so no need to fill in the logs with this message), but it could be handy when debugging if the removal doesn't happen for some reason.

marchello2000 commented on orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/providers/aws/cloudformation/DeleteCloudFormationChangeSetTask.java:
```diff
@@ -0,0 +1,79 @@
+    } else {
+      log.debug(
+          "Not deleting CloudFormation ChangeSet {} from stack {}.", changeSetName, stackName);
+      return TaskResult.builder(ExecutionStatus.SKIPPED).build();
```
Currently, `DeleteCloudFormationChangeSetTask` is the last task in the stage, so returning `SKIPPED` is fine, but if you ever add other tasks after it, if you return `SKIPPED` subsequent tasks won't run (see https://github.com/spinnaker/orca/blob/master/orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/CompleteTaskHandler.kt#L71).

So I recommend either returning `SUCCEEDED` or `FAILED_CONTINUE`

xavileon replied:
ah well spotted. We plan to add more tasks, but I don't think we want to impose the restriction of this one being the last. So I'll move it to `SUCCEEDED` even though it won't actually do any side effect.

Follow-up maybe as another feature/PR: Would it make sense to have a task result that is `NOOP` or the likes? (so we can show the user that that specific task didn't do any action).

marchello2000 commented on orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/providers/aws/cloudformation/WaitForCloudFormationCompletionTask.java:
```diff
@@ -72,8 +72,15 @@ public TaskResult execute(@Nonnull Stage stage) {
-      if (isComplete(status) || isEmptyChangeSet(stage, stack)) {
+      if (isComplete(status)) {
         return TaskResult.builder(ExecutionStatus.SUCCEEDED).outputs(stack).build();
+      } else if (isEmptyChangeSet(stage, stack)) {
```
Again, apologies - I don't know much about CF... I assume, but want to confirm: Will `isComplete` return false in case of an empty changeset? Or should the `if isEmptyChangeSet...` be before `if isCompelte`?

xavileon replied:
yes it will return false as an empty changeset shows as `FAILED` in cloud formation terms. It definitely needs to go before the `isFailed` though. Maybe this part is calling for a refactor in a later PR.

marchello2000 commented on orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/providers/aws/cloudformation/WaitForCloudFormationCompletionTask.java:
```diff
@@ -72,8 +72,15 @@ public TaskResult execute(@Nonnull Stage stage) {
         return TaskResult.builder(ExecutionStatus.SUCCEEDED).outputs(stack).build();
+      } else if (isEmptyChangeSet(stage, stack)) {
+        String changeSetName = (String) result.get("changeSetName");
+        log.debug("CloudFormation ChangeSet {} empty. Requesting to be deleted.", changeSetName);
```
`debug` or `info`? again, just calling it out to make sure it's appropriate level - I think most installation have `debug` logging turned off

xavileon replied:
I'll change this one to info as it is taking an action 👍

marchello2000 commented on orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/providers/aws/DeployCloudFormationStage.java:
```diff
@@ -36,5 +38,8 @@ public void taskGraph(@Nonnull Stage stage, @Nonnull TaskNode.Builder builder) {
         .withTask("forceRefreshCache", CloudFormationForceCacheRefreshTask.class)
         .withTask("waitForCloudFormationCompletion", WaitForCloudFormationCompletionTask.class);
+    if ((boolean) Optional.ofNullable(stage.getContext().get("isChangeSet")).orElse(false)) {
+      builder.withTask("deleteCloudFormationChangeSet", DeleteCloudFormationChangeSetTask.class);
```
do you need to have some sort of `waitForCompletion` / `monitorKatoTask` after the deletion of the changeset?

xavileon replied:
Added the monitor kato task. I'd like to keep it simple for now (no retries). I was looking at the saga backed operations, but this looks like a bigger change and understanding better how it works.

xavileon commented:
Addressed comments in https://github.com/spinnaker/orca/pull/3234/commits/49e608a21c1d70288f0dc0d62e9091b95a8675ae Also adding a monitor kato task to react with a failed operation in clouddriver.
